python/sglang/srt/constrained/grammar.py

```diff
@@ -140,9 +139,52 @@ def fill_vocab_mask(self, vocab_mask: torch.Tensor, vocab_size: int):
             vocab_mask[guide.get_next_instruction(state).tokens] = 0
 
 
+class Grammar:
+    data: Union[Future, GrammarInner]
+
+    def __init__(self, data: Future) -> None:
+        self.data = data
+
+    def _get(self) -> GrammarInner:
+        if isinstance(self.data, Future):
+            self.data = self.data.result()
+        assert isinstance(self.data, GrammarInner)
+        return self.data
+
+    def is_complete(self) -> bool:
+        if isinstance(self.data, Future):
+            return self.data.done()
+        return True
+
+    def accept_token(self, token: int):
+        self._get().accept_token(token)
+
+    def try_jump(self, tokenizer) -> JumpHelper:
+        return self._get().try_jump(tokenizer)
+
+    def jump_forward_str_state(self, helper: JumpHelper) -> Tuple[str, int]:
+        return self._get().jump_forward_str_state(helper)
+
+    def jump_and_retokenize(
+        self, old_output_ids: List[int], new_output_ids: List[int], next_state: int
+    ):
+        self._get().jump_and_retokenize(old_output_ids, new_output_ids, next_state)
+
+    def fill_vocab_mask(self, vocab_mask: torch.Tensor, vocab_size: int):
+        self._get().fill_vocab_mask(vocab_mask, vocab_size)
+
+    # forward all the function calls to the inner object
+    # def __getattr__(self, name):
+    #     if isinstance(self.data, FutureObject):
+    #         self.data = self.data.get()
+    #     assert isinstance(self.data, GrammarInner)
+    #     return getattr(self.data, name)
+
+
 class GrammarCache:
     grammar_cache: Union[BNFCache, FSMCache]
     jump_cache: Union[XGrammarJump, JumpForwardCache, None]
+    executor: ThreadPoolExecutor
```

Comment:
The same, either make `GrammarCache` a dataclass or move these fields to `__init__`